Autocomplete: Improve menu behavior

danielchalmers opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->

Follows up on #8758:
New API, Refactor, Prevent race conditions, Fix menu flashing, Update tests.
This is a must-have for v7-preview1 because of the menu flashing bug in the docs on Firefox.

### Goal
To make the Autocomplete's menu more stable, extensible, and predictable.

### Improvements to API & Behaviour
- Replaces `ChangeMenuAsync` with clearly defined public `OpenMenuAsync` and `CloseMenuAsync`
- Adds classes to input adornment icon and clear button for simpler targeting
- Renames `SelectOnClick` to `SelectOnActivation` for clarity
- Modifies KeyUp/KeyDown handlers to have protected access instead of internal
- Enables holding arrow keys to navigate between menu items (KeyDown instead of KeyUp)

### Resolved Issues
- Prevented multiple menu openings on repeated clicks during search
- Fixed menu flashing issue in Firefox caused by `onclick` and `onfocus` conflict
- Resolves #6961

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

unit

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [X] I've added relevant tests.

danielchalmers committed:
wip

danielchalmers committed:
wip

danielchalmers commented:
@ScarletKuro Could you try your scenarios with this PR and help me with the remaining tests?

danielchalmers committed:
don't focus after selecting an item

danielchalmers committed:
Merge SearchAndOpenMenuAsync into OpenMenuAsync

danielchalmers committed:
wip

danielchalmers committed:
wip

danielchalmers committed:
wip

danielchalmers committed:
fix text selection

danielchalmers committed:
open if clicked while already focused

danielchalmers committed:
Merge branch 'dev' into autocomplete-race-condition

henon commented:
Please set a breakpoint and find out what the exception and the stacktrace are and post it here.

danielchalmers commented:
@henon

> ''<' is an invalid start of a value. Path: $ | LineNumber: 0 | BytePositionInLine: 0.'
> 
> '   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)\n   at System.Text.Json.Serialization.JsonConverter\`1[[System.Collections.Generic.List\`1[[MudBlazor.Examples.Data.Models.Element, MudBlazor.Examples.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)\n   at System.Text.Jso…xample.Search(String value, CancellationToken token) in C:\\Users\\danie\\source\\repos\\MudBlazor\\src\\MudBlazor.Docs\\Pages\\Components\\Autocomplete\\Examples\\AutocompleteStrictFalseExample.razor:line 25\n   at MudBlazor.MudAutocomplete`1.<OpenMenuAsync>d__188[[MudBlazor.Examples.Data.Models.Element, MudBlazor.Examples.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext() in C:\\Users\\danie\\source\\repos\\MudBlazor\\src\\MudBlazor\\Components\\Autocomplete\\MudAutocomplete.razor.cs:line 649'

ScarletKuro commented:
> @henon @ScarletKuro Why does the SearchFunc throw an exception here

Sorry, but I really don't know. I have no codebase knowledge on how this component functions at slightest. @henon @mckaragoz has more experience with this component.
It's also hard for me to adequately review this, you changed many places and this component code resembles delicate paper, susceptible to crumbling with the slightest disturbance :flushed: Because it has many race conditions and my async void -> Task change is a proof of this. I appropriate your effort on trying to fix it tho, but almost feels like it would be better to make a parallel component called "MudAutocompleteX" and do minimum at least for the mudblazor docs search, battle test it there and then extend the functionality.

danielchalmers commented:
@ScarletKuro I think I've definitely hardened the code and made it less brittle and susceptible to race conditions. It does need some heavy testing though. I would hate to discard these changes from the BCL because it has a lot of improvements for all users.

danielchalmers committed:
fix focus after long search

henon commented:
The exception means that the json coming from the web api is invalid. For whatever reason. I can't see why your changes would cause that

henon commented:
Maybe you can find out what the api call returns that can't be parsed?

Yomodo commented:
> Maybe you can find out what the api call returns that can't be parsed?

Probably the result is returned as plain text (server error) while json is expected.

henon commented:
Does it even work without your changes?

danielchalmers commented:
> Does it even work without your changes?

@henon Works on https://dev.mudblazor.com/components/autocomplete#strict-mode

jperson2000 commented:
> @henon
> 
> > ''<' is an invalid start of a value. Path: $ | LineNumber: 0 | BytePositionInLine: 0.'
> > '   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)\n   at System.Text.Json.Serialization.JsonConverter`1[[System.Collections.Generic.List`1[[MudBlazor.Examples.Data.Models.Element, MudBlazor.Examples.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)\n   at System.Text.Jso…xample.Search(String value, CancellationToken token) in C:\Users\danie\source\repos\MudBlazor\src\MudBlazor.Docs\Pages\Components\Autocomplete\Examples\AutocompleteStrictFalseExample.razor:line 25\n   at MudBlazor.MudAutocomplete`1.d__188[[MudBlazor.Examples.Data.Models.Element, MudBlazor.Examples.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext() in C:\Users\danie\source\repos\MudBlazor\src\MudBlazor\Components\Autocomplete\MudAutocomplete.razor.cs:line 649'

@danielchalmers I've seen this error before when I expect JSON but get HTML, such as when I forget to authenticate an HttpClient request.   If we can get the raw response from the `HttpClient`, then we can look at the actual response content and dig deeper.  Here's some temporary code which should do the trick:

```
  private async Task<IEnumerable<Element>> Search(string value, CancellationToken token)
    {
        // return await httpClient.GetFromJsonAsync<List<Element>>($"webapi/periodictable/{value}", token);
        var response = await httpClient.GetAsync($"webapi/periodictable/{value}", token);
        var content = await response.Content.ReadAsStringAsync();
        Debugger.Break(); // Examine the content variable here for issues, maybe an auth error?
        return JsonSerializer.Deserialize<List<Element>>(content);
    }
```

danielchalmers committed:
Fix IsOpen logic

danielchalmers committed:
fix test logic

danielchalmers commented:
@jperson2000 Thanks for the great tip, it made me discover that the webapi returns 404, so I guess it's not available while debugging WASM and was never a regression @henon 

The rest was just some weird logic that should be fixed and ready to review now.

danielchalmers committed:
change open logic for #6961

henon commented:
> > Does it even work without your changes?
> 
> @henon Works on https://dev.mudblazor.com/components/autocomplete#strict-mode

Yes, but try to execute it locally instead from dev.mudblazor.ocm. It may well be the cause of the error.

ScarletKuro commented:
> Thanks for the great tip, it made me discover that the webapi returns 404

Are you starting docs as wasm standalone? There is WasmHost then the API should be available.

henon commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -725,54 +741,35 @@ private string GetItemString(T item)
         }
 
-        internal virtual async Task OnInputKeyDown(KeyboardEventArgs args)
+        protected virtual async Task OnInputKeyDownAsync(KeyboardEventArgs args)
```
`protected` makes it `public` for derived components. I don't know why this is even virtual. We could make it just `private` instead.

henon commented:
The `Strict` parameter and example troubles me. I can't think of a use-case where you want the `Strict="true"` behavior because it will show an empty dropdown when clicking the autocomplete which is just nonsense. I remember that this behavior was the default in the beginning. We added the parameter to avoid a breaking change because users wanted the `Strict="false"` behavior instead.

Unless somebody can think of a valid use-case for `Strict="true"` I would now consider taking the opportunity and making the breaking change of just removing this parameter.

mikes-gh commented:
> > Thanks for the great tip, it made me discover that the webapi returns 404
> 
> Are you starting docs as wasm standalone? There is WasmHost then the API should be available.

Yes this.  wasm standalone doesnt have any mvc controllers to provide the endpoints for the examples data.

danielchalmers committed:
change accessibility

danielchalmers committed:
Docs: Remove `Strict` mode

henon commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -1154,44 +1105,6 @@ public async Task Autocomplete_Should_Raise_KeyDown_KeyUp_Event()
         }
 
-        /// <summary>
-        /// Test case for <seealso cref="https://github.com/MudBlazor/MudBlazor/issues/6412"/>
-        /// </summary>
-        [Test]
-        public async Task Autocomplete_Should_Highlight_Selected_Item_After_Disabled()
-        {
-            var disabledItemSelector = "mud-list-item-disabled";
-            var selectedItemSelector = "mud-selected-item";
-            var popoverSelector = "div.mud-popover";
-
-            var selectedItemString = "peach";
-            var disabledItemString = "carrot";
-
-            var comp = Context.RenderComponent<AutocompleteStrictFalseSelectedHighlight>();
-            var autocompletecomp = comp.FindComponent<MudAutocomplete<string>>();
-            var autocomplete = autocompletecomp.Instance;
-
-            // Select the peach list item
-            autocompletecomp.Find("input").Input(selectedItemString);
-            comp.WaitForAssertion(() => comp.Find(popoverSelector).ClassList.Should().Contain("mud-popover-open"));
-            await comp.InvokeAsync(async () => await autocompletecomp.Find("input").KeyUpAsync(new KeyboardEventArgs() { Key = "Enter" }));
-            autocomplete.Text.Should().Be(selectedItemString);
-            autocomplete.Value.Should().Be(selectedItemString);
-
-            // Opening the list of autocomplete
-            await comp.InvokeAsync(autocompletecomp.Instance.OpenMenuAsync);
-            comp.WaitForAssertion(() => comp.Find(popoverSelector).ClassList.Should().Contain("mud-popover-open"));
-            var listItems = comp.FindComponents<MudListItem<string>>().ToArray();
-
-            // Ensure that the carrot list item is disabled
-            var disabledItem = listItems.Single(x => x.Markup.Contains(disabledItemSelector));
-            disabledItem.Markup.Should().Contain(disabledItemString);
-
-            // Assert if the peach is highlighted
-            var selectedItem = listItems.Single(x => x.Markup.Contains(selectedItemSelector));
-            selectedItem.Markup.Should().Contain(selectedItemString);
-        }
-
```
revert pls

henon commented on src/MudBlazor.UnitTests/Components/AutocompleteTests.cs:
```diff
@@ -1066,55 +1066,6 @@ public async Task Autocomplete_Should_HaveValueWithTextChangedEvent()
-
-            var comp = Context.RenderComponent<AutocompleteStrictFalseTest>();
-            var autocompletecomp = comp.FindComponents<MudAutocomplete<AutocompleteStrictFalseTest.State>>()[index];
-            var autocomplete = autocompletecomp.Instance;
-
-            //search for and select California
-            autocompletecomp.Find("input").Input("Calif");
-            comp.WaitForAssertion(() => comp.FindAll("div.mud-popover")[index].ClassList.Should().Contain("mud-popover-open"));
-            await comp.InvokeAsync(async () => await autocompletecomp.Find("input").KeyUpAsync(new KeyboardEventArgs() { Key = "Enter" }));
-            autocomplete.Text.Should().Be(californiaString);
-            autocomplete.Value.StateName.Should().Be(californiaString);
-
-            //California should appear as index 5 and be selected
-            await comp.InvokeAsync(autocompletecomp.Instance.OpenMenuAsync); // reopen menu because Enter closes it.
-            comp.WaitForAssertion(() => comp.FindAll("div.mud-popover")[index].ClassList.Should().Contain("mud-popover-open"));
-            var items = comp.FindComponents<MudListItem<AutocompleteStrictFalseTest.State>>().ToArray();
-            items.Length.Should().Be(10);
-            var item = items.SingleOrDefault(x => x.Markup.Contains(californiaString));
-            items.ToList().IndexOf(item).Should().Be(5);
-            comp.WaitForAssertion(() => items.Single(s => s.Markup.Contains(californiaString)).Find(listItemQuerySelector).ClassList.Should().Contain(selectedItemClassName));
-
-            await comp.InvokeAsync(async () => await autocompletecomp.Find("input").KeyUpAsync(new KeyboardEventArgs() { Key = "Escape" })); // Close autocomplete.
-
-            //search for and select Virginia
-            autocompletecomp.Find("input").Input("Virginia");
-            comp.WaitForAssertion(() => comp.FindAll("div.mud-popover")[index].ClassList.Should().Contain("mud-popover-open"));
-            await comp.InvokeAsync(async () => await autocompletecomp.Find("input").KeyUpAsync(new KeyboardEventArgs() { Key = "Enter" }));
-            autocomplete.Text.Should().Be(virginiaString);
-            autocomplete.Value.StateName.Should().Be(virginiaString);
-
-            //West Virginia is not in the first 10 states, so it should not appear in the list
-            await comp.InvokeAsync(autocompletecomp.Instance.OpenMenuAsync); // reopen menu because Enter closes it.
-            comp.WaitForAssertion(() => comp.FindAll("div.mud-popover")[index].ClassList.Should().Contain("mud-popover-open"));
-            var items2 = comp.FindComponents<MudListItem<AutocompleteStrictFalseTest.State>>().ToArray();
-            items2.Length.Should().Be(10);
-            var item2 = items2.SingleOrDefault(x => x.Markup.Contains(virginiaString));
-            items2.ToList().IndexOf(item).Should().Be(-1);
-            items2.Count(s => s.Find(listItemQuerySelector).ClassList.Contains(selectedItemClassName)).Should().Be(0);
-        }
-
```
revert pls

henon commented on src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs:
```diff
@@ -671,14 +655,7 @@ public async Task OpenMenuAsync()
-            else
-            {
-                _selectedListItemIndex = _enabledItemIndices.Any() ? _enabledItemIndices[0] : -1;
-            }
```
I think this will have to be reverted as well because `searchingWhileSelected` was true when Strict was false which should be default behavior

danielchalmers committed:
Revert "Docs: Remove `Strict` mode"

This reverts commit f63501b252209bfe15f903b0247f2ebd61c84c22.

Mr-Technician commented:
@danielchalmers We still want to retain `Strict==false`, we just want to remove the undesired `Strict==true` behavior that results in empty lists.

henon reviewed (approved):
The changes look good. I don't see any red flags. Of course, you can never know with an unstable component like this which was in a delicate balance before. Let's merge and hope for the best. If this introduces issues we can always revert.

danielchalmers committed:
dont blur on escape

not typical behavior and you can just use tab now

kind of stops you from closing dialogs with double escape though

danielchalmers commented:
@Mr-Technician @ScarletKuro Looks good?
